### PR TITLE
8328723: IP Address error when client enables HTTPS endpoint check on server socket

### PR DESCRIPTION
**Pull request by pkumaraswamy**
The client identity checks when "HTTPS" endpoint identification algorithm is set on SSL server throws "java.security.cert.CertificateException: No subject alternative names present" when client certificate's SubjectAltName extension does not match its IP address

Since the server has no external knowledge of what the client's identity ought to be,  HTTPS identity checks must be disabled on the server side.
The exception message has been fixed to indicate the same.

I have performed the test both on SSL Server Engine and SSL Server Socket and attached are logs and snapshot for reference, also I have ran the changes against external test suite and test runs are green.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328723](https://bugs.openjdk.org/browse/JDK-8328723): IP Address error when client enables HTTPS endpoint check on server socket (**Bug** - P3)

### Reviewers
 * [Bradford Wetmore](https://openjdk.org/census#wetmore) (@bradfordwetmore - **Reviewer**)
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20048/head:pull/20048` \
`$ git checkout pull/20048`

Update a local copy of the PR: \
`$ git checkout pull/20048` \
`$ git pull https://git.openjdk.org/jdk.git pull/20048/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20048`

View PR using the GUI difftool: \
`$ git pr show -t 20048`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20048.diff">https://git.openjdk.org/jdk/pull/20048.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20048#issuecomment-2210489586)

### CODE REVIEW

**Comment by pkumaraswamy** [comment]
Teast data
[https_endpt.zip](https://github.com/user-attachments/files/16109176/https_endpt.zip)

File: src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java [hunk 1]
```diff
@@ -428,8 +428,17 @@ static void checkIdentity(SSLSession session,
         }
 
         if (!identifiable) {
-            checkIdentity(peerHost,
-                    trustedChain[0], algorithm, chainsToPublicCA);
+            try {
+                checkIdentity(peerHost,
+                        trustedChain[0], algorithm, chainsToPublicCA);
+            } catch(CertificateException ce) {
+                if(checkClientTrusted && "HTTPS".equalsIgnoreCase(algorithm)) {
```
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java:435]: I'm not so sure this is the best fix. It seems that this should be checked earlier when the `setEndpointIdentification` method is called with a value that doesn't make sense for a client. This is not specific to your fix, but I wonder if we should go through all of the effort to create a TLS session and validate the certificate chain only to have it fail at this point. We already knew this was going to be a problem before we started the TLS session. @bradfordwetmore - can you please also weigh in on this fix?

**Comment by seanjmullan** [comment]
I have a general question about this fix. Does it ever make sense for a client to call `SSLParameters.setEndpointIdentification` with a value of "HTTPS"? If not, then it probably makes sense to detect this error situation earlier and throw an Exception, preferably before the TLS handshake even starts. @bradfordwetmore Any thoughts on this one?

File: src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java [hunk 2]
```diff
@@ -428,8 +428,17 @@ static void checkIdentity(SSLSession session,
         }
 
         if (!identifiable) {
-            checkIdentity(peerHost,
-                    trustedChain[0], algorithm, chainsToPublicCA);
+            try {
```
Comment (bradfordwetmore) [inline src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java:431]: Minor formatting nits/suggestions:
```
if(
->
if (
```

Lines <= 80 chars, please.

```
Endpoint Identification algorithm
->
Endpoint Identification Algorithm
```

File: src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java [hunk 3]
```diff
@@ -428,8 +428,17 @@ static void checkIdentity(SSLSession session,
...
+                checkIdentity(peerHost,
```
Comment (bradfordwetmore) [inline src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java:432]: You might do the direct check for the `checkClientTrusted`/`HTTPS` and immediately throw.  If that doesn't fail, then do the fallback `checkIdentity`.  Doing so would save the overhead of calling `checkIdentity` only to find out there was an error.

```
if (!identifiable) {
    // Clients can't identify themselves via SNI/hostnames in HTTPS.
    if (checkClientTrusted && "HTTPS".equalsIgnoreCase(algorithm)) {
        throw...
    }
    checkIdentity(peerHost, trustedChain[0], algorithm, chainsToPublicCA);
}
```

**Review by bradfordwetmore — COMMENTED** [review]
To @seanjmullan 's question.  

`SSLParameters` can be set to anything before the handshake starts, even potentially to conflicting values like this.  Only once handshaking is underway can we determine the inconsistency and report the error.

See code for additional comments, then I can approve.

File: src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java (hunk 3 above)
Comment (pkumaraswamy) [reply src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java:432]: Initially I had the same thought, however if an existing app has a ceritficate with it's IP address entry in SubjectAltName extension and endpoint set to HTTPS it is expected handshake will pass without any issues.

If I do this check first, then it'll break existing application where certificate has been created to match this criteria.

File: src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java (hunk 2 above)
Comment (pkumaraswamy) [reply src/java.base/share/classes/sun/security/ssl/X509TrustManagerImpl.java:431]: Done, attached is the test result 
[https_endpt.zip](https://github.com/user-attachments/files/16139516/https_endpt.zip)

**Review by bradfordwetmore — APPROVED** [review]
LGTM

**Review by djelinski — APPROVED** [review]

**Pull request closed without merging**